### DISCUSSION TITLE
Install packages from Github, Gitlab and Gitea instances

### DISCUSSION THREAD

**Discussion by trymeouteh** [composer/composer#11228] [General]
Github is a popular git service and Gitlab and Gitea allows you to self host your own git servers. It would be amazing to see the ability to install packages directly from any github repo and any gitlab or gitea instance.

And to have the ability to have packages use dependancies directly from Github, Gitlab and Gitea repository and not have to only rely on dependencies from the Packagist repository.

Such a feature will allow for decentralisation, private repositories and the ability to truly self host your own packages.

**Comment by stof** [comment]
Composer already supports that:

- metadata can come from several repositories, not just from packagist.org (repo.packagist.org is registered by default, but it can also be disabled). It is still recommended to aggregate your metadata using Private Packagist, Satis or another solution exposing a metadata repository in the composer format as this is a lot faster than reading metadata from each VCS repository each time.
- authentication is supported for repositories, which means private repositories are totally possible (the composer authentication system is keyed by domain name, so if you have several private repositories needing different credentials, they cannot be deployed at subfolders of the same domain, but they should use different domains instead)
- Composer supports configuring some domains as being gitlab domains, which allows it to use the gitlab features (APIs to read metadata instead of cloning the git repo, using dist archives generated by gitlab, etc...), which allows to have good support for self-hosted gitlab instances (composer has similar support for GitHub Entreprise Server by configuring the github domains btw)

Composer does not yet have native support for Gitea to be able to read archives generated by Gitea (if it does) so it would fallback to its generic git support which requires cloning the repo to install the package (often much slower than downloading only the code corresponding to a tag).

**Reply by trymeouteh** [reply]
When I run this command in composer for example to install moment from the github repository (not packagist). I get an error message

`composer require https://github.com/moment/moment`

How do you install packages from github?

**Reply by Seldaek** [reply]
You read the docs ;) https://getcomposer.org/doc/05-repositories.md#vcs
